identity-account-storage/src/types/key_location.rs

```diff
@@ -5,61 +5,163 @@ use core::fmt::Debug;
 pub struct KeyLocation {
-  method: MethodType,
-  fragment: Fragment,
-  generation: Generation,
+  /// The [`KeyType`] of the key.
+  pub key_type: KeyType,
+  /// The fragment of the key.
+  fragment: String,
+  /// The hash of the public key.
+  #[serde(with = "key_hash_serialization")]
+  pub(in crate::types::key_location) key_hash: u64,
 }
 
 impl KeyLocation {
-  /// Creates a new `KeyLocation`.
-  pub fn new(method: MethodType, fragment: String, generation: Generation) -> Self {
-    Self {
-      method,
-      fragment: Fragment::new(fragment),
+  /// Create a location from a [`KeyType`], the fragment of a verification method
+  /// and the bytes of a public key.
+  pub fn new(key_type: KeyType, fragment: String, public_key: &[u8]) -> Self {
+    let mut hasher = SeaHasher::new();
+    hasher.write(public_key);
+    let key_hash: u64 = hasher.finish();
 
-      generation,
+    Self {
+      key_type,
+      fragment,
+      key_hash,
     }
   }
 
-  /// Returns the method type of the key location.
-  pub fn method(&self) -> MethodType {
-    self.method
-  }
+  /// Obtain the location of a verification method's key in storage.
+  pub fn from_verification_method(method: &IotaVerificationMethod) -> crate::Result<Self> {
+    let fragment: &str = method
+      .id()
+      .fragment()
+      .ok_or(crate::Error::DIDError(identity_did::Error::MissingIdFragment))?;
+    let method_data: &MethodData = method.data();
 
-  /// Returns the fragment name of the key location.
-  pub fn fragment(&self) -> &Fragment {
-    &self.fragment
-  }
+    let key_type: KeyType = match method.type_() {
+      MethodType::Ed25519VerificationKey2018 => KeyType::Ed25519,
+      MethodType::X25519KeyAgreementKey2019 => KeyType::X25519,
+    };
+
+    let public_key: Vec<u8> = method_data.try_decode()?;
 
-  /// Returns the fragment name of the key location.
-  pub fn fragment_name(&self) -> &str {
-    self.fragment.name()
+    Ok(KeyLocation::new(key_type, fragment.to_owned(), public_key.as_ref()))
   }
 
-  /// Returns the integration generation when this key was created.
-  pub fn generation(&self) -> Generation {
-    self.generation
+  /// Returns the canonical string representation of the location.
+  ///
+  /// This should be used as the representation for storage keys.
+  pub fn canonical_repr(&self) -> String {
+    format!("{}:{}", self.fragment, self.key_hash)
   }
 }
 
 impl Display for KeyLocation {
   fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-    f.write_fmt(format_args!("({}:{}:{})", self.generation, self.fragment, self.method))
+    f.write_str(&self.canonical_repr())
   }
 }
 
-impl Debug for KeyLocation {
-  fn fmt(&self, f: &mut Formatter<'_>) -> Result {
-    f.write_fmt(format_args!("KeyLocation{}", self))
+// Custom Hash and Equality implementations to not include the key_type.
+
+impl Hash for KeyLocation {
+  fn hash<H: Hasher>(&self, state: &mut H) {
+    self.fragment.hash(state);
+    self.key_hash.hash(state);
+  }
+}
```

Comment:
I feel like this should use `canonical_repr` internally. This likely has the same platform-dependent bug with `hash`.

Reply:
Good point, although the hash is not exposed to other languages right now (could be in the future I suppose), and only used so we can use `KeyLocation` as the key in `MemStore` rather than converting to strings.

Reply:
Even used internally I think it may break in Wasm, similar to `KeyLocation::new`.

Reply:
True that, I updated it to use `KeyLocation::canonical`.